GEODE-4650: Refactor Elder selection

balesh2 opened the pull request:
* add distributed test for elder
* Extract the ClusterElderManager from ClusterDistributionManager and
delegate
* add unit test for ClusterElderManager
* test how elder is chosen to reproduce GEODE-4650
* remove caching of elder. This avoids an issue where in some cases, the
cached value could be set to a member that has left the cluster, causing
the DLockService to hang.
* remove the tryLock used to retry in getElderState. tryLock was
leftover and unnecessary after removing CollaborationLock.

Signed-off-by: Dan Smith <dsmith@pivotal.io>
Signed-off-by: Galen O'Sullivan <gosullivan@pivotal.io>

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

dschneider-pivotal commented on geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterElderManager.java:
```diff
@@ -0,0 +1,208 @@
+    } finally {
+      this.elderLock.unlock();
+    }
+    this.elderStateInitialized = true;
```
Having this line here is okay, but I think it would be even better if it was moved into the "if (this.elderState == null)" block. That way we only set it to true when we are the thread setting "elderState".

dschneider-pivotal commented on geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterElderManager.java:
```diff
@@ -0,0 +1,208 @@
+      waitForElder(clusterDistributionManager.getId());
+    }
+
+    if (clusterDistributionManager.getId().equals(getElderCandidate())) {
```
looks like you can change this line to: "if (isElder())"

dschneider-pivotal commented on geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterElderManager.java:
```diff
@@ -0,0 +1,208 @@
+    }
+  }
+
+  private ElderState initializeElderState() {
```
would a better name be "getElderStateInitializingIfNeeded"?

dschneider-pivotal commented on geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterElderManager.java:
```diff
@@ -0,0 +1,208 @@
+
+    if (clusterDistributionManager.getId().equals(getElderCandidate())) {
+      // we are the elder
+      if (this.elderStateInitialized) {
```
I think it would be easier to understand if this block, "if (this.elderStateInitialized {...}" was moved into the initializeElderState method.

balesh2 commented on geode-core/src/distributedTest/java/org/apache/geode/distributed/internal/locks/DLockServiceCharacterizationTests.java:
```diff
@@ -34,19 +44,27 @@
 import org.apache.geode.internal.cache.DistributedRegion;
+import org.apache.geode.test.junit.rules.ConcurrencyRule;
 
 public class DLockServiceCharacterizationTests {
```
This test does get run as a DistributedTest.

balesh2 commented:
@dschneider-pivotal We do want to move the log out of the loop. Now that we've changed how that looping is happening, we were getting that log message every 100ms, so it wasn't useful. I've added the check before logging or getting elder candidates

pdxrunner commented on geode-core/src/distributedTest/java/org/apache/geode/distributed/internal/locks/DLockServiceCharacterizationTests.java:
```diff
@@ -16,15 +16,25 @@
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertSame;
```
Again, a mixture of assertion libraries for no apparent reason other than new tests use the preferred assertj, but older tests not updated.

Oops, some of the new tests don't use assertj

pdxrunner commented on geode-dunit/src/main/java/org/apache/geode/test/junit/rules/VMProvider.java:
```diff
@@ -82,6 +82,15 @@ public void invoke(String name, final SerializableRunnableIF runnable) {
+  }
+
+
+  public <T> AsyncInvocation<T> invokeAsync(String name, final SerializableCallableIF<T> runnable) {
```
Change parameter name `runnable` -> `callable`

balesh2 commented on geode-core/src/distributedTest/java/org/apache/geode/distributed/DistributedLockServiceDUnitTest.java:
```diff
@@ -395,132 +263,118 @@ public void testOneGetsAndOthersTimeOut() throws Exception {
     doOneGetsAndOthersTimeOut(4, 3);
```
✅ done

pdxrunner commented on geode-core/src/distributedTest/java/org/apache/geode/distributed/DistributedLockServiceDUnitTest.java:
```diff
@@ -1639,8 +1707,10 @@ public void run() {
-  public void testBug32461() throws Exception {
-    logger.debug("[testBug32461] prepping");
+  public void testBug32461() {
+    if (logger.isDebugEnabled()) {
```
My understanding is that this test is done in product code for performance - `logger.debug()` has a higher overhead then testing `logger.isDebugEnabled()`. So, using this conditional in tests is not usually necessary.

balesh2 replied:
Thanks for the info! I just assumed that it was needed based on Darrel's comment but good to know that I don't have to do that in the future.

galen-pivotal commented:
Could you please add something to the commit message explaining why it was safe to remove `isAdam()` and replace it with `isLoner()`?

balesh2 commented:
[Draft commit message for when we squash this]
GEODE-4650: Resolve race condition in selection of the elder

* no longer cache the elder, re-compute the elder when needed
* extract elder logic to a new class to make unit testing possible
* adds tests for elder selection
* adds tests of DLock Grantor failover
* removes isAdam() - isAdam used to mean that the member was alone (that there were no non-surprise, non-admin members in the cluster) when it joined. This was only used in two places. The first, in the DLockService, protected against recovering dlocks when there isn't a cluster. This usage is replaced with a check for isLoner(). The other use of isAdam was in ElderInitProcessor and was redundant with an inner check if there were other members in the distributed system.
* fix testFairness so that it can be run repeatedly in the same JVM
